addons/binding/org.openhab.binding.energenie/src/main/java/org/openhab/binding/energenie/handler/EnergenieGatewayHandler.java

```diff
@@ -0,0 +1,619 @@
+import org.openhab.binding.energenie.internal.api.manager.FailingRequestHandler;
+import org.openhab.binding.energenie.internal.api.manager.FailingRequestHandlerImpl;
+import org.openhab.binding.energenie.internal.api.manager.EnergenieApiConfiguration;
+import org.openhab.binding.energenie.internal.api.manager.EnergenieApiManager;
+import org.openhab.binding.energenie.internal.api.manager.EnergenieApiManagerImpl;
+import org.openhab.binding.energenie.internal.api.manager.ThingCallback;
+import org.openhab.binding.energenie.internal.rest.RestClient;
+import org.osgi.framework.ServiceReference;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.gson.JsonArray;
+import com.google.gson.JsonElement;
+import com.google.gson.JsonObject;
+
+/**
+ * Handler for the Mi|Home Gateway (MIHO001). The gateway is a central
+ * communication hub between the individual devices.
+ *
+ * @author Mihaela Memova - Initial contribution
+ * @author Svilen Valkanov - Added methods to get information about subdevices
+ */
+public class EnergenieGatewayHandler extends BaseBridgeHandler {
+
+    // Warning messages used when some unexpected behavior is encountered
+    private static final String UNREGISTERED_GATEWAY_MESSAGE = "The gateway has been unregistered";
+    private static final String ALREADY_REGISTERED_GATEWAY_MESSAGE = "The gateway has already been registered";
+    private static final String NOT_SUCCESSFULL_REGISTRATION_MESSAGE = "Registering new gateway was not successfull";
+    private static final String NOT_ACTIVE_GATEWAY_MESSAGE = "The gateway has not been active in the last two minutes. Please check your connection.";
+    private static final String CHANGING_LABEL_MESSAGE = "Mi|Home REST API does not allow to change gateway's label. Changing the thing's label will not affect the device label.";
+    private static final String INVALID_USER_MESSAGE = "Invalid username. Email address expected.";
+    private static final String INVALID_GATEWAY_CODE_MESSAGE = "Invalid gateway code. 10 capital letters and numbers expected";
+    private static final String REJECTED_EXECUTION_OF_UPDATE_TASK_MESSAGE = "The update task cannot be scheduled for execution";
+    private static final String UNSUCCESSFUL_DATE_PARSING_MESSAGE = "Could not get the 'inactive' time interval of the gateway. Something went wrong with the parsing.";
+    private static final int INITIAL_REFRESH_DELAY_SECONDS = 10;
+
+    /**
+     * The longest allowed inactive time interval (in seconds) of the gateway
+     * device. After that interval, if the gateway is still inactive, the thing
+     * goes offline. If a gateway is active or not is determined by its
+     * 'last_seen' property.
+     */
+    private static final int MAXIMUM_INACTIVE_PERIOD_SECONDS = 120;
+
+    /** Pattern used to check if a given String represents an email address */
+    public static final String EMAIL_PATTERN = ".+.*@.+.*(\\.[-A-Za-z]{2,})";
```

Comment:
In general it would be wiser to store the compiled regex as a static and re-use that, in this case you only use it few times so just keep it in mind.